packages/cardano-services/src/Program/services/pgboss.ts

```diff
@@ -0,0 +1,163 @@
+
+          await dataSource.initialize();
+
+          return dataSource;
+        })()
+      )
+    )
+  );
+
+export interface PgBossServiceConfig {
+  parallelJobs: number;
+  queues: PgBossQueue[];
+}
+
+export interface PgBossServiceDependencies {
+  connectionConfig$: Observable<PgConnectionConfig>;
+  db: Pool;
+  logger: Logger;
+}
+
+export class PgBossHttpService extends HttpService {
+  #config: PgBossServiceConfig;
+  #connectionConfig$: Observable<PgConnectionConfig>;
+  #dataSource$: ReplaySubject<DataSource | null> = new ReplaySubject(1);
+  #dataSourceSubscription: Subscription;
+  #firstSubscription = true;
+  #pgBoss: PgBoss;
+
+  constructor(cfg: PgBossServiceConfig, deps: PgBossServiceDependencies) {
+    const { connectionConfig$, db, logger } = deps;
+    const provider: Provider = { healthCheck: () => Promise.resolve({ ok: false }) };
+
+    super('pg-boss-service', provider, Router(), logger);
+
+    provider.healthCheck = async () => {
+      try {
+        await db.query('SELECT 1');
+      } catch (error) {
+        this.logger.error(error);
+
+        return { ok: false };
+      }
+
+      return { ok: true };
+    };
+
+    this.#connectionConfig$ = connectionConfig$;
+    this.#config = cfg;
+    this.#pgBoss = createPgBoss(db, this.logger);
+  }
+
+  private createWorkerDataSource() {
+    // The first time we subscribe we want to immediately do that;
+    // the next time it is after an error, so we want to add a short delay
+    this.#dataSourceSubscription = createWorkerDataSource(
+      this.#firstSubscription ? this.#connectionConfig$ : timer(500).pipe(mergeMap(() => this.#connectionConfig$)),
+      this.logger
+    ).subscribe(this.#dataSource$);
+
+    this.#firstSubscription = false;
+  }
+
+  private recreateWorkerDataSource() {
+    this.#dataSource$.next(null);
+    this.#dataSourceSubscription.unsubscribe();
+    this.createWorkerDataSource();
+  }
+
+  getDataSource() {
+    return firstValueFrom(this.#dataSource$.pipe(filter(isNotNil)));
+  }
+
+  protected async initializeImpl() {
+    this.createWorkerDataSource();
+    await this.#pgBoss.start();
+  }
+
+  protected async startImpl() {
+    for (const queue of this.#config.queues) {
+      const createHandler = queueHandlers[queue];
+      const logger = contextLogger(this.logger, queue);
+      const handler = createHandler(this.#dataSource$.pipe(filter(isNotNil)), logger);
+
+      await this.#pgBoss.work(
+        queue,
+        {
+          teamConcurrency: this.#config.parallelJobs,
+          teamRefill: true,
+          teamSize: this.#config.parallelJobs
+        },
+        async (job) => {
+          const { id, data } = job;
+
+          try {
+            logger.info(`Job ${id} started`, data);
+            await handler(data);
+            logger.info(`Job ${id} successfully completed`, data);
+          } catch (error) {
+            // In case of error, pg-boss doesn't log anything and silently retry the job later
```

Comment:
```suggestion
            // pg-boss may retry the job eventually, but this is opaque, so the error is logged here for more insight.
```

Can remove "may" if the retry is always expected

Reply:
The retry options depend on the job queue so "may fits" here.
65ef5f92e should solve this